File: experiments/gilded-rose/python/attempts/opus46high/1/gilded_rose.py
def _clamp_quality(quality):
    return max(0, min(50, quality))


def _update_normal(item):
    degradation = 1 if item.sell_in > 0 else 2
    item.quality = _clamp_quality(item.quality - degradation)
    item.sell_in -= 1


def _update_aged_brie(item):
    increase = 1 if item.sell_in > 0 else 2
    item.quality = _clamp_quality(item.quality + increase)
    item.sell_in -= 1


def _update_backstage_passes(item):
    if item.sell_in <= 0:
        item.quality = 0
    elif item.sell_in <= 5:
        item.quality = _clamp_quality(item.quality + 3)
    elif item.sell_in <= 10:
        item.quality = _clamp_quality(item.quality + 2)
    else:
        item.quality = _clamp_quality(item.quality + 1)
    item.sell_in -= 1


def _update_sulfuras(item):
    pass


def _update_conjured(item):
    degradation = 2 if item.sell_in > 0 else 4
    item.quality = _clamp_quality(item.quality - degradation)
    item.sell_in -= 1


_UPDATERS = {
    "Aged Brie": _update_aged_brie,
    "Backstage passes to a TAFKAL80ETC concert": _update_backstage_passes,
    "Sulfuras, Hand of Ragnaros": _update_sulfuras,
    "Conjured Mana Cake": _update_conjured,
}


class GildedRose:
    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            updater = _UPDATERS.get(item.name, _update_normal)
            updater(item)


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)

File: experiments/gilded-rose/js/attempts/opus46high/1/gilded-rose.js
export class Item {
  constructor(name, sellIn, quality) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

function clampQuality(quality) {
  return Math.max(0, Math.min(50, quality));
}

function updateNormal(item) {
  const degradation = item.sellIn <= 0 ? 2 : 1;
  item.quality = clampQuality(item.quality - degradation);
  item.sellIn -= 1;
}

function updateAgedBrie(item) {
  const increase = item.sellIn <= 0 ? 2 : 1;
  item.quality = clampQuality(item.quality + increase);
  item.sellIn -= 1;
}

function updateBackstagePass(item) {
  if (item.sellIn <= 0) {
    item.quality = 0;
  } else if (item.sellIn <= 5) {
    item.quality = clampQuality(item.quality + 3);
  } else if (item.sellIn <= 10) {
    item.quality = clampQuality(item.quality + 2);
  } else {
    item.quality = clampQuality(item.quality + 1);
  }
  item.sellIn -= 1;
}

function updateSulfuras() { }

function updateConjured(item) {
  const degradation = item.sellIn <= 0 ? 4 : 2;
  item.quality = clampQuality(item.quality - degradation);
  item.sellIn -= 1;
}

function getUpdater(name) {
  if (name === 'Aged Brie') return updateAgedBrie;
  if (name === 'Backstage passes to a TAFKAL80ETC concert') return updateBackstagePass;
  if (name === 'Sulfuras, Hand of Ragnaros') return updateSulfuras;
  if (name.startsWith('Conjured')) return updateConjured;
  return updateNormal;
}

export class Shop {
  constructor(items = []) {
    this.items = items;
  }

  updateQuality() {
    for (const item of this.items) {
      getUpdater(item.name)(item);
    }
    return this.items;
  }
}
